### PR TITLE
Add http streaming capability

### PR DESCRIPTION
**Pull request by m4tt075**
An attempt to "resurrect" @FernetMenta 's PR #471.

### CODE REVIEW

**Comment by m4tt075** [comment]
Please don't let emotions stand in the way of good ideas and progress...

@ksooo, @Jalle19 : I've added the proposed https addition in an extra commit, but - just like @FernetMenta - have no option to test it. As you have requested the addition, could you please verify it is in fact working for https as well?

File: pvr.hts/addon.xml.in [hunk 1]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <addon
   id="pvr.hts"
-  version="7.1.4"
+  version="7.2.4"
```
Comment (ksooo) [inline pvr.hts/addon.xml.in]: 7.2.0

File: pvr.hts/changelog.txt [hunk 2]
```diff
@@ -1,3 +1,6 @@
+7.2.4
```
Comment (ksooo) [inline pvr.hts/changelog.txt]: 7.2.0

File: src/Tvheadend.cpp [hunk 3]
```diff
@@ -367,6 +367,34 @@ PVR_ERROR CTvheadend::GetChannels(bool radio, kodi::addon::PVRChannelsResultSet&
...
+
+  if (!Settings::GetInstance().GetStreamingHTTP())
+    return PVR_ERROR_NO_ERROR;
+
+  auto it = m_channels.find(channel.GetUniqueId());
+  if (it == m_channels.end())
+    return PVR_ERROR_FAILED;
+
+  std::string url = Settings::GetInstance().GetUseHTTPS() ? "https://" : "http://";
+  url += Settings::GetInstance().GetUsername() + ":" + Settings::GetInstance().GetPassword() + "@";
```
Comment (ksooo) [inline src/Tvheadend.cpp]: user name and/or password might be empty, as @Jalle19 already explained.

File: src/Tvheadend.h [hunk 4]
```diff
@@ -91,6 +91,7 @@ class ATTRIBUTE_HIDDEN CTvheadend : public kodi::addon::CInstancePVRClient,
 
   PVR_ERROR GetChannelsAmount(int& amount) override;
   PVR_ERROR GetChannels(bool radio, kodi::addon::PVRChannelsResultSet& results) override;
+  PVR_ERROR GetChannelStreamProperties(const kodi::addon::PVRChannel& channel, std::vector<kodi::addon::PVRStreamProperty>& properties) override;
```
Comment (ksooo) [inline src/Tvheadend.h]: clang-format!

**Review by ksooo — COMMENTED** [review]
Needs a bit more love and please squash 1st and second commit into one.

File: pvr.hts/resources/language/resource.language.en_gb/strings.po [hunk 5]
```diff
@@ -459,7 +459,11 @@ msgctxt "#30504"
 msgid "Stream read chunk size for recordings (KB)"
 msgstr ""
 
-#empty strings from id 30505 to 30509
+msgctxt "#30505"
+msgid "Use http streaming for channels"
```
Comment (Jalle19) [inline pvr.hts/resources/language/resource.language.en_gb/strings.po]: HTTP

File: pvr.hts/addon.xml.in (hunk 1 above)
Comment (m4tt075) [reply pvr.hts/addon.xml.in]: Done.

File: src/Tvheadend.cpp (hunk 3 above)
Comment (m4tt075) [reply src/Tvheadend.cpp]: Please review whether I've addressed @Jalle19's comment correctly.

**Comment by m4tt075** [comment]
Many thanks for reviewing! I've just force pushed the requested changes, but have not yet had a chance to runtime-test them. Could you point me at a Win64 build, please?

File: src/Tvheadend.cpp [hunk 6]
```diff
@@ -367,6 +367,39 @@ PVR_ERROR CTvheadend::GetChannels(bool radio, kodi::addon::PVRChannelsResultSet&
...
+
+  auto it = m_channels.find(channel.GetUniqueId());
+  if (it == m_channels.end())
+    return PVR_ERROR_FAILED;
+
+  std::string url = Settings::GetInstance().GetUseHTTPS() ? "https://" : "http://";
+  std::string usn = Settings::GetInstance().GetUsername();
+  std::string pas = Settings::GetInstance().GetPassword();
+
+  if (usn[0] && pas[0])
```
Comment (Jalle19) [inline src/Tvheadend.cpp]: I think empty() is the C++ way to check this

Comment (m4tt075) [reply src/Tvheadend.cpp]: Ooops.

**Review by Jalle19 — APPROVED** [review]

**Comment by Jalle19** [comment]
Seems okay to me

**Comment by ksooo** [comment]
Please squash commit 2 and 3.

File: src/Tvheadend.cpp [hunk 7]
```diff
@@ -367,6 +367,39 @@ PVR_ERROR CTvheadend::GetChannels(bool radio, kodi::addon::PVRChannelsResultSet&
...
+
+  if (!usn.empty() && !pas.empty())
+    url += usn + ":" + pas + "@";
+
+  url += Settings::GetInstance().GetHostname() + ":";
+  url += std::to_string(Settings::GetInstance().GetPortHTTP());
+  url += "/stream/channelid/";
+  url += std::to_string(it->first);
+
+  std::string streamingProfile = Settings::GetInstance().GetStreamingProfile();
```
Comment (ksooo) [inline src/Tvheadend.cpp]: `const std::string`, please.

File: src/Tvheadend.cpp [hunk 8]
```diff
@@ -367,6 +367,39 @@ PVR_ERROR CTvheadend::GetChannels(bool radio, kodi::addon::PVRChannelsResultSet&
   return PVR_ERROR_NO_ERROR;
 }
 
+PVR_ERROR CTvheadend::GetChannelStreamProperties(const kodi::addon::PVRChannel& channel, std::vector<kodi::addon::PVRStreamProperty>& properties)
+{
+  CLockObject lock(m_mutex);
```
Comment (ksooo) [inline src/Tvheadend.cpp]: Move obtaining the lock just before line `auto it = m_channels.find(channel.GetUniqueId());`.

File: src/Tvheadend.cpp [hunk 9]
```diff
@@ -367,6 +367,39 @@ PVR_ERROR CTvheadend::GetChannels(bool radio, kodi::addon::PVRChannelsResultSet&
...
+
+  if (!Settings::GetInstance().GetStreamingHTTP())
+    return PVR_ERROR_NO_ERROR;
+
+  auto it = m_channels.find(channel.GetUniqueId());
+  if (it == m_channels.end())
```
Comment (ksooo) [inline src/Tvheadend.cpp:378]: As  `it` is only used once, there's no need for a variable. 
=> `if (m_channels.find(channel.GetUniqueId() == m_channels.end())`

File: src/Tvheadend.cpp [hunk 10]
```diff
@@ -367,6 +367,39 @@ PVR_ERROR CTvheadend::GetChannels(bool radio, kodi::addon::PVRChannelsResultSet&
...
+    return PVR_ERROR_FAILED;
+
+  std::string url = Settings::GetInstance().GetUseHTTPS() ? "https://" : "http://";
+  std::string usn = Settings::GetInstance().GetUsername();
```
Comment (ksooo) [inline src/Tvheadend.cpp]: `const std::string`

File: src/Tvheadend.cpp [hunk 11]
```diff
@@ -367,6 +367,39 @@ PVR_ERROR CTvheadend::GetChannels(bool radio, kodi::addon::PVRChannelsResultSet&
...
+  std::string pas = Settings::GetInstance().GetPassword();
```
Comment (ksooo) [inline src/Tvheadend.cpp]: `const std::string`

File: pvr.hts/changelog.txt [hunk 12]
```diff
@@ -1,3 +1,6 @@
+7.2.0
+- Add capability to use http streaming instead of htsp
```
Comment (ksooo) [inline pvr.hts/changelog.txt]: HTTP, HTSP

**Review by ksooo — COMMENTED** [review]
Minors only.

**Comment by ksooo** [comment]
I just remembered that we have https://github.com/kodi-pvr/pvr.hts/blob/Matrix/src/tvheadend/HTSPConnection.cpp#L130 ff., which contains everything that has been added in this PR to build the URL (plus support for webroot, which is missing in this PR). We should just use the existing method.

File: src/Tvheadend.cpp (hunk 9 above)
Comment (m4tt075) [reply src/Tvheadend.cpp:378]: `it` is used again in line 391.

**Comment by m4tt075** [comment]
@ksooo Thanks for your comments. I've addressed them so far. I'm just not sure how to call `getweburl` correctly in that context (my cpp "skills" are VERY basic). Any advice appreciated. Will squash my commits as soon as you are happy with the code.

File: src/Tvheadend.cpp (hunk 9 above)
Comment (ksooo) [reply src/Tvheadend.cpp:378]: Ah, right.

File: src/Tvheadend.cpp [hunk 13]
```diff
@@ -367,6 +367,39 @@ PVR_ERROR CTvheadend::GetChannels(bool radio, kodi::addon::PVRChannelsResultSet&
...
+  std::string url = Settings::GetInstance().GetUseHTTPS() ? "https://" : "http://";
+  const std::string usn = Settings::GetInstance().GetUsername();
+  const std::string pas = Settings::GetInstance().GetPassword();
+
+  if (!usn.empty() && !pas.empty())
+    url += usn + ":" + pas + "@";
+
+  url += Settings::GetInstance().GetHostname() + ":";
+  url += std::to_string(Settings::GetInstance().GetPortHTTP());
+  url += "/stream/channelid/";
```
Comment (ksooo) [inline src/Tvheadend.cpp]: No need to calculate the basic parts of the URL yourself. We have already have a helper method for that.

```c++
std::string path = "/stream/channelid/" + std::to_string(it->first);
const std::string streamingProfile = Settings::GetInstance().GetStreamingProfile();
if (!streamingProfile.empty())
  path += "?profile=" + streamingProfile;

const std::string url = m_conn.GetWebURL("%s", path);
```

**Review by ksooo — CHANGES_REQUESTED** [review]

File: src/Tvheadend.cpp (hunk 13 above)
Comment (m4tt075) [reply src/Tvheadend.cpp]: Thanks for your patience! I understand now.

The build failed after I applied the change suggesting to change the last line to
```cpp
const std::string url = m_conn->GetWebURL("%s", path);
``` 
Even with that amendment the builds still fail though. I'm getting:
```cpp
/home/travis/build/kodi-pvr/pvr.hts/src/Tvheadend.cpp:386:51: error: cannot pass
      object of non-trivial type 'std::string' (aka 'basic_string<char>')
      through variadic method; call will abort at runtime [-Wnon-pod-varargs]
  const std::string url = m_conn->GetWebURL("%s", path);
```
I assume I need to `static_cast` something, but again, out of my depths, here. Sorry for the hassle.

Comment (ksooo) [reply src/Tvheadend.cpp]: path.c_str()

Comment (m4tt075) [reply src/Tvheadend.cpp]: Thanks for the c++ crash course, @ksooo. Working again...  :-)

**Comment by m4tt075** [comment]
OK. Squashed all my fixup commits...

**Comment by ksooo** [comment]
OK, have you runtime-tested the latest version of this PR?

**Comment by m4tt075** [comment]
No. Could you point me at a test build for Win64, please? And please note that I won't be able to test the https pathway. Therefore my earlier remark.
@hagaygo, @cvh: I know you are interested in this, too. Any additional tests appreciated.

**Comment by ksooo** [comment]
Test builds: https://jenkins.kodi.tv/view/Addons/job/kodi-pvr/job/pvr.hts/view/change-requests/job/PR-476/

File: src/tvheadend/Settings.h [hunk 14]
```diff
@@ -146,6 +149,7 @@ class Settings
   void SetAutorecApproxTime(int value) { m_iAutorecApproxTime = value; }
   void SetAutorecMaxDiff(int value) { m_iAutorecMaxDiff = value; }
   void SetStreamingProfile(const std::string& value) { m_strStreamingProfile = value; }
+  void SetStreamingHTTP(const bool& value) { m_bUseHTTPStreaming = value; }
```
Comment (ksooo) [inline src/tvheadend/Settings.h]: Oh, just spotted this one. please change to `void SetStreamingHTTP(bool value) { m_bUseHTTPStreaming = value; }`

**Review by ksooo — CHANGES_REQUESTED** [review]
One more thing.

**Comment by ksooo** [comment]
Please squash and confirm successful runtime test and this one is ready to go in.

**Review by ksooo — APPROVED** [review]
Looks good to me now.

**Comment by hagaygo** [comment]
> 
> 
> Test builds: https://jenkins.kodi.tv/view/Addons/job/kodi-pvr/job/pvr.hts/view/change-requests/job/PR-476/

Hi, 

Just tested it on matrix master X64 and it seems to works great , tvheadend also reports "pass" profile and playback seems to be working great on my problematic TS's. 👍 

One downside as expected is that timeshift is not available.

2 Questions

1.Any chance this will get backported to Leia branch ? so we can have this feature on current kodi versions ?
2.If timeshift is not gonna be implmented using http streaming (too complicated/not possible ?) is it possible to add option to override streaming profile (htsp/http) on channel level ? (even without GUI support , just a manual XML edit would do 😃  )

**Comment by ksooo** [comment]
@hagaygo thanks for testing.

>> Any chance this will get backported to Leia branch ?

 Sorry, no. If you need this feature go with Kodi 19 Alpha, which imo is already pretty stable. I'm using the nighlies for months in my living room without any problems.

>> is it possible to add option to override streaming profile (htsp/http) on channel level ?

 Everything is possible if somebody steps up and opens a PR for this. Personally, I will not find the time for this, unfortunately.

**Comment by m4tt075** [comment]
@ksooo Wow, that was quick now. I've just confirmed that - AFTER your last change request last night - it works properly for http streaming (https still untested). But you probably knew that anyway!?   :-)   In any case, my sincere thanks to you and @Jalle19 for reengaging and "carrying me over the line" here... 

And many, many thanks to @FernetMenta for the idea and implementation in the first place. Please do keep sharing them! They are truly appreciated!!!

**Comment by CvH** [comment]
After trying a bit around, are 4-6sec for a single channel switch normal ?
With hts I have a round 1-2sec.

**Comment by m4tt075** [comment]
I've had around 10 seconds with the original version. With the merged version I'm down to 3-4 sec, but, yes, tuning is slower than htsp for me, too. Once the stream has been tuned it runs rock solid though.

**Comment by FernetMenta** [comment]
You should report this issue to Kodi. Tuning streams like this was used to be much faster. Most likely it got broken with a recent ffmpeg upgrade. This also affects other pvr addons.

**Comment by hagaygo** [comment]
Did you try on latest nightly ? 

When I tried on x64 build it seemed fine, I will double check it later today.

Just checked , HTTP Channel switching is kinda the same as HTSP.

**Comment by ksooo** [comment]
Can you please shift this discussion to the Kodi forum? Every comment here triggers a lot of people and the discussion is not really related to this PR. Thanks. Or open an issue for Kodi to discuss this.

**Comment by FernetMenta** [comment]
I consider it very related to this PR. People might think that this new feature has the downside of longer tuning times per se. But this is not the case. It can be as fast as with hts.

**Comment by m4tt075** [comment]
I can confirm your assumptions: Whilst http streaming took longer to tune than htsp streaming on Kodi alpha-2 (my previous test system), it tunes just as quickly on today's nightly (KodiSetup-20201018-6384782d-master-x64). As such, the problem is indeed not related to this PR, and apparently already addressed on the Kodi side...

**Comment by disaster123** [comment]
Is it still true that http mode does not support timeshift? Here seems to be a commit to enable timeshift: https://github.com/jahutchi/pvr.hts/commit/f463ee16032625b86a3467cbc16dfdaffcc8b90a

Is it still true that htsp is outdated and should be replaced by http?

**Comment by ksooo** [comment]
True and false.

**Pull request merged**